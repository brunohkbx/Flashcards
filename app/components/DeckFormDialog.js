import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import * as Yup from 'yup';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  HelperText,
  TextInput
} from 'react-native-paper';

export const Form = props => {
  const {
    visible,
    handleDismiss,
    handleSubmit,
    deck,
    title,
    actionSubmitText
  } = props;

  validationSchema = Yup.object().shape({
    title: Yup.string().required('Required')
  });

  return (
    <Formik
      initialValues={deck}
      validationSchema={validationSchema}
      enableReinitialize
      render={props =>
        <DeckFormDialog
          visible={visible}
          handleDismiss={() => { handleDismiss(); props.resetForm({}); }}
          title={title}
          actionSubmitText={actionSubmitText}
          {...props}
        />
      }
      onSubmit={(values, actions) => handleSubmit(values, actions)}
    />
  )
}

export class DeckFormDialog extends Component {
  render() {
    const {
      visible,
      handleDismiss,
      handleSubmit,
      handleChange,
      handleBlur,
      touched,
      errors,
      values,
      title,
      actionSubmitText
    } = this.props;

    return (
      <Dialog
        visible={visible}
        onDismiss={handleDismiss}
      >
        <DialogTitle>{title}</DialogTitle>
        <DialogContent>
          <TextInput
            label="Title"
            value={values.title}
            onChangeText={handleChange('title')}
            onBlur={handleBlur('title')}
            error={!!(touched.title && errors.title)}
            autoFocus
          />
          <HelperText type="error" visible={!!(touched.title && errors.title)}>
            {errors.title}
          </HelperText>
        </DialogContent>
        <DialogActions>
          <Button primary onPress={handleDismiss}>Cancel</Button>
          <Button primary onPress={handleSubmit}>{actionSubmitText}</Button>
        </DialogActions>
      </Dialog>
    );
  }
}

Form.propTypes = {
  visible: PropTypes.bool,
  handleDismiss: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  actionSubmitText: PropTypes.string.isRequired
}

Form.defaultProps = {
  visible: false
}

export default Form;
