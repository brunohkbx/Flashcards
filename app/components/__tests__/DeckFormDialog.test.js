import React from 'react';
import { shallow } from 'enzyme';
import { DeckFormDialog, Form } from '../DeckFormDialog';

describe('Dumb DeckFormDialog', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      visible: true,
      handleDismiss: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(() => jest.fn),
      handleBlur: jest.fn(() => jest.fn),
      touched: {},
      errors: {},
      values: {}
    }, propOverrides)

    const wrapper = shallow(<DeckFormDialog {...defaultProps} />);

    const cancel = () => {
      wrapper
        .find('DialogActions')
        .childAt(0)
        .simulate('press')
    }

    const submit = () => {
      wrapper
        .find('DialogActions')
        .childAt(1)
        .simulate('press')
    }

    const dismiss = () => {
      wrapper.simulate('dismiss');
    }

    return {
      defaultProps,
      wrapper,
      cancel,
      submit,
      dismiss
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  test('onDismiss calls handleDismiss', () => {
    const mockHandleDismiss = jest.fn();
    const { dismiss } = setup({ handleDismiss: mockHandleDismiss });

    dismiss();

    expect(mockHandleDismiss).toHaveBeenCalled();
  });

  describe('Cancel button', () => {
    it('calls handleDismiss', () => {
      const mockHandleDismiss = jest.fn();
      const { cancel } = setup({ handleDismiss: mockHandleDismiss });

      cancel();

      expect(mockHandleDismiss).toHaveBeenCalled();
    });
  });

  describe('Submit button', () => {
    it('calls handleSubmit', () => {
      const mockHandleSubmit = jest.fn();
      const { submit } = setup({ handleSubmit: mockHandleSubmit });

      submit();

      expect(mockHandleSubmit).toHaveBeenCalled();
    });

    it('displays an error message if input is invalid', () => {
      const { wrapper } = setup({
        touched: { title: true },
        errors: { title: 'Required' }
      });

      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('DeckFormDialog enhanced with Formik ', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      visible: true,
      handleDismiss: jest.fn(),
      handleSubmit: jest.fn(),
      createDeck: jest.fn(),
    }, propOverrides)

    const wrapper = shallow(<Form {...defaultProps} />);

    const submit = () => {
      wrapper.simulate(
        'submit',
        {},
        Object.assign({ resetForm: jest.fn() }, propOverrides)
      )
    }

    const dismiss = () => {
      wrapper
        .dive()
        .find('DeckFormDialog')
        .shallow()
        .simulate('dismiss')
    }

    return {
      defaultProps,
      wrapper,
      submit,
      dismiss
    }
  }

  describe('Cancel button ', () => {
    it('dismisses the form', () => {
      const mockHandleDismiss = jest.fn();
      const { dismiss } = setup({ handleDismiss: mockHandleDismiss });

      dismiss();

      expect(mockHandleDismiss).toHaveBeenCalled();
    });
  });

  describe('Submit button', () => {
    it('submits the form', () => {
      const mockHandleSubmit = jest.fn();
      const { submit } = setup({ handleSubmit: mockHandleSubmit });

      submit();

      expect(mockHandleSubmit).toHaveBeenCalled();
    });
  });
});
