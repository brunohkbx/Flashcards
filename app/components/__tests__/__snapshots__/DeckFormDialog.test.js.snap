// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dumb DeckFormDialog Submit button displays an error message if input is invalid 1`] = `
<withTheme(Dialog)
  onDismiss={[MockFunction]}
  visible={true}
>
  <withTheme(DialogTitle)>
    Create New Deck
  </withTheme(DialogTitle)>
  <DialogContent>
    <withTheme(TextInput)
      autoFocus={true}
      error={true}
      label="Title"
      onBlur={[Function]}
      onChangeText={[Function]}
    />
    <withTheme(HelperText)
      type="error"
      visible={true}
    >
      Required
    </withTheme(HelperText)>
  </DialogContent>
  <DialogActions>
    <withTheme(Button)
      onPress={[MockFunction]}
      primary={true}
    >
      Cancel
    </withTheme(Button)>
    <withTheme(Button)
      onPress={[MockFunction]}
      primary={true}
    >
      Create
    </withTheme(Button)>
  </DialogActions>
</withTheme(Dialog)>
`;

exports[`Dumb DeckFormDialog renders properly 1`] = `
<withTheme(Dialog)
  onDismiss={[MockFunction]}
  visible={true}
>
  <withTheme(DialogTitle)>
    Create New Deck
  </withTheme(DialogTitle)>
  <DialogContent>
    <withTheme(TextInput)
      autoFocus={true}
      error={false}
      label="Title"
      onBlur={[Function]}
      onChangeText={[Function]}
    />
    <withTheme(HelperText)
      type="error"
      visible={false}
    />
  </DialogContent>
  <DialogActions>
    <withTheme(Button)
      onPress={[MockFunction]}
      primary={true}
    >
      Cancel
    </withTheme(Button)>
    <withTheme(Button)
      onPress={[MockFunction]}
      primary={true}
    >
      Create
    </withTheme(Button)>
  </DialogActions>
</withTheme(Dialog)>
`;
